File: apps/driver-android/src/main/kotlin/lk/mageride/driver/push/PushRouter.kt
package lk.mageride.driver.push

import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.channels.BufferOverflow
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.SharedFlow
import kotlinx.coroutines.flow.asSharedFlow
import lk.mageride.driver.nav.DriverRoute

/**
 * One notification as notification-svc sends it.
 *
 * The shape is C051's `data` dictionary, not a guess: the service writes `kind`, `deeplink` and
 * `notificationId` on every push it produces, plus per-kind extras (`rideId`, `fare`, `distance`
 * on an offer; `requestId`, `ttl` on a location request). The C051 handoff spells out the client
 * half — *"the `data.kind` switch (`ride_offer`, `location_request`, and the type name for
 * everything else), and `data.deeplink`"*.
 *
 * Everything is a `String`: FCM's data payload is `Map<String, String>` on the wire and coercing
 * here would throw inside a broadcast receiver, which is the one place a crash is invisible.
 *
 * @property kind `data.kind`. Lower-case for the three the service special-cases, the SCREAMING
 *   type name otherwise.
 * @property deeplink `data.deeplink` — a `mageride://…` URI, absent on kinds that open nothing.
 * @property notificationId `data.notificationId`, the id `POST /v1/notify/ack` takes.
 * @property data The whole dictionary, for a screen that needs a per-kind extra.
 */
internal data class PushMessage(
    val kind: String?,
    val deeplink: String?,
    val notificationId: String?,
    val data: Map<String, String>,
) {
    internal companion object {
        /** Reads one off an FCM `RemoteMessage.getData()`. */
        fun from(data: Map<String, String>): PushMessage = PushMessage(
            kind = data[KEY_KIND],
            deeplink = data[KEY_DEEPLINK],
            notificationId = data[KEY_NOTIFICATION_ID],
            data = data,
        )

        const val KEY_KIND: String = "kind"
        const val KEY_DEEPLINK: String = "deeplink"
        const val KEY_NOTIFICATION_ID: String = "notificationId"

        /** E-01's 15 s atomic offer. The only kind that takes over the screen. */
        const val KIND_RIDE_OFFER: String = "ride_offer"
    }
}

/**
 * Turns a push into a destination, and hands it to whoever is showing the UI.
 *
 * **A deep link is resolved, never trusted.** `deeplink` arrives over the network; mapping it to a
 * known [DriverRoute] rather than handing the raw URI to the navigator means an unrecognised or
 * hostile value opens Home instead of whatever a future `mageride://` handler might accept.
 *
 * The router is a process singleton in the Koin graph because both entry points feed it: the FCM
 * service while the app is backgrounded, and `MainActivity` when a notification tap starts the
 * app with an intent. Its flow replays one value, so a push that arrives before the shell is
 * composed is still delivered — the alternative is a tapped notification that silently does
 * nothing on a cold start.
 */
internal class PushRouter {

    private val mutablePending = MutableSharedFlow<DriverRoute>(
        replay = 1,
        extraBufferCapacity = 1,
        onBufferOverflow = BufferOverflow.DROP_OLDEST,
    )

    /** Destinations a push asked for. Collected once, by the shell. */
    val pending: SharedFlow<DriverRoute> = mutablePending.asSharedFlow()

    /** Offers [message]'s destination to the shell. No-op when the push opens nothing. */
    fun offer(message: PushMessage) {
        routeFor(message)?.let(mutablePending::tryEmit)
    }

    /** Offers a raw `mageride://…` URI — the notification-tap path, where only the intent survives. */
    fun offer(uri: String?) {
        resolve(uri)?.let(mutablePending::tryEmit)
    }

    /**
     * Forgets the replayed value once the shell has navigated, so a rotation does not re-navigate.
     *
     * `resetReplayCache` is the one API that expresses "this was consumed" on a replaying
     * `SharedFlow`; the alternative is a nullable wrapper every collector has to unwrap.
     */
    @OptIn(ExperimentalCoroutinesApi::class)
    fun consume() {
        mutablePending.resetReplayCache()
    }

    internal companion object {

        /** The destination [message] should open. */
        fun routeFor(message: PushMessage): DriverRoute? = when (message.kind) {
            // The offer sheet is not a deep link — it is a takeover the dashboard owns (C070),
            // and it must open even though `offer.created` carries a ride deeplink for the
            // passenger side. Routing it to Home is what puts the driver where the sheet shows.
            PushMessage.KIND_RIDE_OFFER -> DriverRoute.Home

            else -> resolve(message.deeplink)
        }

        /**
         * `mageride://ride/{id}` → the active-ride screen, and so on for the four links
         * notification-svc mints (`DeepLinks` in its `EventHandlers.cs`; C051 note (n)).
         *
         * Anything else — a scheme that is not ours, a host with no screen, a malformed URI —
         * is `null`. Parsing by hand rather than with `android.net.Uri` keeps this testable on
         * the JVM, which is where the table is asserted.
         */
        @Suppress("ReturnCount") // One guard per malformed-input case; nesting them reads worse.
        fun resolve(uri: String?): DriverRoute? {
            val value = uri?.trim().orEmpty()
            if (!value.startsWith(SCHEME)) return null

            val path = value.removePrefix(SCHEME).trim('/')
            if (path.isEmpty()) return null

            val segments = path.split('/')
            val id = segments.getOrNull(1)?.takeIf(String::isNotBlank)
            return when (segments[0]) {
                HOST_RIDE -> id?.let(DriverRoute::ActiveRide)
                HOST_PACKAGE -> id?.let(DriverRoute::ActiveRide)
                HOST_WALLET -> DriverRoute.Wallet
                HOST_DOCUMENTS -> DriverRoute.Documents
                else -> null
            }
        }

        /** D6' I-23.3's scheme. */
        const val SCHEME: String = "mageride://"

        private const val HOST_RIDE = "ride"

        // A package delivery IS a ride (R-01 keeps one aggregate), so both links land on the
        // same screen. The distinction the passenger app makes — SCR-PA-021 — has no driver-side
        // counterpart: a driver carrying a parcel is on the ordinary active-ride screen.
        private const val HOST_PACKAGE = "package"
        private const val HOST_WALLET = "wallet"
        private const val HOST_DOCUMENTS = "documents"
    }
}

File: apps/driver-ios/DriverApp/Push/PushRouter.swift
import Combine
import Foundation

/// One notification as notification-svc sends it.
///
/// The shape is C051's `data` dictionary, not a guess: the service writes `kind`, `deeplink` and
/// `notificationId` on every push it produces, plus per-kind extras (`rideId`, `fare`, `distance`
/// on an offer; `requestId`, `ttl` on a location request).
///
/// Everything is a `String`: an APNs payload is JSON, but notification-svc's `data` block is the
/// same dictionary FCM carries as `Map<String, String>` on Android, and coercing a type here would
/// throw inside a background delivery handler — the one place a crash is invisible.
struct PushMessage: Equatable {

    /// `data.kind`. Lower-case for the ones the service special-cases, the SCREAMING type name
    /// otherwise.
    let kind: String?

    /// `data.deeplink` — a `mageride://…` URI, absent on kinds that open nothing.
    let deeplink: String?

    /// `data.notificationId`, the id `POST /v1/notify/ack` takes.
    let notificationId: String?

    /// The whole dictionary, for a screen that needs a per-kind extra.
    let data: [String: String]

    /// Reads one off an APNs `userInfo`.
    ///
    /// The keys are read from the top level, which is where notification-svc puts them: APNs has no
    /// `data` envelope of its own, so a custom key sits beside `aps` rather than inside it. Values
    /// that are not strings are dropped rather than described — a numeric `fare` arriving as a JSON
    /// number is a contract question, not something to stringify here and lose.
    static func from(userInfo: [AnyHashable: Any]) -> PushMessage {
        var data: [String: String] = [:]
        for (key, value) in userInfo {
            guard let key = key as? String, let value = value as? String else { continue }
            data[key] = value
        }
        return PushMessage(
            kind: data[Keys.kind],
            deeplink: data[Keys.deeplink],
            notificationId: data[Keys.notificationId],
            data: data
        )
    }

    enum Keys {
        static let kind = "kind"
        static let deeplink = "deeplink"
        static let notificationId = "notificationId"
    }

    /// E-01's 15 s atomic offer. The only kind that takes over the screen.
    static let kindRideOffer = "ride_offer"

    /// US-6A.15's 30-minute reminder before a scheduled pickup (D5' §14.4).
    ///
    /// The SCREAMING form, because notification-svc's catalogue spells it that way
    /// (`NotificationCatalogue.ScheduledReminder`) and `kind` carries the type name for every push
    /// the service does not special-case.
    static let kindScheduledReminder = "SCHEDULED_REMINDER"
}

/// Turns a push into a destination, and hands it to whoever is showing the UI.
///
/// **A deep link is resolved, never trusted.** `deeplink` arrives over the network; mapping it to a
/// known ``DriverRoute`` rather than handing the raw URI to the navigator means an unrecognised or
/// hostile value opens nothing instead of whatever a future `mageride://` handler might accept.
///
/// A process singleton, because both entry points feed it: the notification delegate while the app
/// is backgrounded, and the launch options when a notification tap starts the app cold. ``pending``
/// keeps its last value for the same reason the Android router replays one — a push that arrives
/// before the shell is on screen is still delivered, and the alternative is a tapped notification
/// that silently does nothing on a cold start.
@MainActor
final class PushRouter: ObservableObject {

    /// The destination a push asked for, if it has not been consumed yet. Read once, by the shell.
    ///
    /// Not `private(set)`: the shell subscribes to `$pending`, and a projected value behind a
    /// private setter is the kind of access-control subtlety that changes between Swift releases.
    /// Nothing else writes it — ``offer(_:)`` and ``consume()`` are the whole surface.
    @Published var pending: DriverRoute?

    /// Offers [message]'s destination. No-op when the push opens nothing.
    func offer(_ message: PushMessage) {
        if let route = PushRouter.route(for: message) { pending = route }
    }

    /// Offers a raw `mageride://…` URI — the notification-tap path, where only the link survives.
    func offer(uri: String?) {
        if let route = PushRouter.resolve(uri) { pending = route }
    }

    /// Forgets the pending destination once the shell has navigated, so a scene change does not
    /// navigate a second time.
    func consume() {
        pending = nil
    }

    // MARK: - The tables
    //
    // Both are `static` and pure so `PushRouterTests` can assert them without a running app,
    // which is where the Android side puts them too.

    /// The destination [message] should open.
    static func route(for message: PushMessage) -> DriverRoute? {
        switch message.kind {
        // The offer sheet is not a deep link — it is a takeover the dashboard owns (C088), and it
        // must open even though `offer.created` carries a *ride* deeplink minted for the passenger
        // side. Routing it to Home is what puts the driver where the sheet shows.
        case PushMessage.kindRideOffer:
            return .home

        // US-6A.15 — D2' §SCR-DI-018: "30-min reminder push deep-links here". It carries no
        // deeplink to follow: `DeepLinks` in Notification.Api mints four URIs and none of them
        // names a scheduled ride, so the routing is on the **type**, which D5' §14.4 and
        // notification-svc's catalogue both fix. If a `mageride://scheduled` host is ever minted,
        // ``resolve(_:)`` is where it belongs.
        case PushMessage.kindScheduledReminder:
            return .scheduledRides

        default:
            return resolve(message.deeplink)
        }
    }

    /// `mageride://ride/{id}` → the active-ride screen, and so on for the four links
    /// notification-svc mints (`DeepLinks` in its `EventHandlers.cs`; C051 note (n)).
    ///
    /// Anything else — a scheme that is not ours, a host with no screen, a malformed URI — is
    /// `nil`. Parsed by hand rather than with `URLComponents` so the table is asserted on exactly
    /// the strings the server sends, and so a percent-encoding quirk cannot change what a link
    /// opens.
    static func resolve(_ uri: String?) -> DriverRoute? {
        let value = uri?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        guard value.hasPrefix(scheme) else { return nil }

        let path = String(value.dropFirst(scheme.count))
            .trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        guard !path.isEmpty else { return nil }

        let segments = path.split(separator: "/").map(String.init)
        let id = segments.count > 1 && !segments[1].isEmpty ? segments[1] : nil

        switch segments[0] {
        case hostRide, hostPackage:
            // A package delivery IS a ride (R-01 keeps one aggregate), so both links land on the
            // same screen. The distinction the passenger app makes — SCR-PA-021 — has no
            // driver-side counterpart.
            return id.map { DriverRoute.activeRide(rideId: $0) }
        case hostWallet:
            return .wallet
        case hostDocuments:
            return .documents
        default:
            return nil
        }
    }

    /// D6' I-23.3's scheme.
    static let scheme = "mageride://"

    private static let hostRide = "ride"
    private static let hostPackage = "package"
    private static let hostWallet = "wallet"
    private static let hostDocuments = "documents"
}
